### ISSUE TITLE
'atomic pull' should support pulling images by digest

### ISSUE THREAD

**Issue by miabbott** [projectatomic/atomic#691]
In order to have more feature parity with `docker`, `atomic pull` should be able to retrieve images using the SHA256 ID.

See the following example using `docker`

```
# docker pull fedora@sha256:64a02df6aac27d1200c2572fe4b9949f1970d05f74d367ce4af994ba5dc3669e                                                                                      
Trying to pull repository registry.access.redhat.com/fedora ... 
unknown: Not Found
Trying to pull repository docker.io/library/fedora ... 
sha256:64a02df6aac27d1200c2572fe4b9949f1970d05f74d367ce4af994ba5dc3669e: Pulling from docker.io/library/fedora
2bf01635e2a0: Pull complete 
Digest: sha256:64a02df6aac27d1200c2572fe4b9949f1970d05f74d367ce4af994ba5dc3669e
Status: Downloaded newer image for docker.io/fedora@sha256:64a02df6aac27d1200c2572fe4b9949f1970d05f74d367ce4af994ba5dc3669e
```

**Comment by baude** [comment]
@rhatdan do you know if the @ sign is unique to this kind of image description/pull?

I think we have the logic more or less, just need to figure out how to decompose this kind of input.

**Comment by rhatdan** [comment]
@runcom PTAL

I have a feeling that this could also be used for other pulls.

**Comment by runcom** [comment]
@baude yes, that @ sign is unique to the pull-by-digest in docker so you should be safe when decomposing a reference (either it has a tag with :tag or a digest with @sha256:hexvalue)

@rhatdan what do you mean by this could also be used for other pulls?

**Comment by rhatdan** [comment]
@runcom I am not sure what I meant.  Ignore it.

**Comment by baude** [comment]
Right now skopeo does not support this.  Will put this on hold until it does and then implement it into atomic.

**Comment by rhatdan** [comment]
@runcom is there any movement on skopeo pulling by digest?

**Comment by runcom** [comment]
Why is skopeo not supporting this already? what's missing here?
```
$ ./skopeo copy docker://busybox@sha256:29f5d56d12684887bdfa50dcd29fc31eea4aaf4ad3bec43daf19026a7ce69912 oci:test

Getting image source signatures
Copying blob sha256:56bec22e355981d8ba0878c6c2f23b21f422f30ab0aba188b54f1ffeff59c190
 603.20 KB / 652.49 KB [==================================================>----]
Copying config sha256:e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba
 0 B / 1.43 KB [---------------------------------------------------------------]
Writing manifest to image destination
Storing signatures
```

**Comment by rhatdan** [comment]
@baude Looks like it is supported?

**Comment by baude** [comment]
I'm working on a PR for this.  One issue we need to decide on is what the tag should be for the image.  Consider what docker does:
```
[bbaude@bbaude atomic (pull_by_digest)]$ docker pull docker.io/busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Trying to pull repository docker.io/library/busybox ... 
sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e: Pulling from docker.io/library/busybox
4b0bc1c4050b: Pull complete 
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for docker.io/busybox@sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
[bbaude@bbaude atomic (pull_by_digest)]$ docker images | grep busy
docker.io/busybox                                         <none>              7968321274dc        2 weeks ago         1.11 MB
[bbaude@bbaude atomic (pull_by_digest)]$
```
Note how <none> is used for the tag.  

Skopeo will not accept <none> as a valid tag.  So should skopeo be fixed to allow this and we replicate docker's behaviour or should we decide on a different tag?

**Comment by runcom** [comment]
@baude `<none>` _is not_ used for the tag. That's just a placeholder to say "no tag". Skopeo should not accept that at all. If you pull by digest you don't get a tag and that's it afaict

**Comment by runcom** [comment]
@mtrmac could you also take a look at this?

**Comment by mtrmac** [comment]
The thing is, `docker load` can not record the manifest digest (that manifest is not stored in the tarball). It would be possible to use `docker load` without a tag, but that image would then be a bit difficult to reference, (it would be referenceable by the `config.json` digest = image ID, but the `config.json` may be manufactured on-the-fly by `skopeo` when converting from schema1). I don’t know, perhaps there is a way to record the digest reference after-the-fact, I haven’t looked at this.

One thing that can certainly be done is to support the `busybox:sometag@sha256:somedigest` syntax, pulling using the digest and using that tag. However that’s probably not quite `docker pull`-compatible either..

**Comment by baude** [comment]
https://github.com/projectatomic/atomic/pull/856

**Comment by baude** [comment]
@miabbott good to go?

**Comment by miabbott** [comment]
Fixed by #856
